libfido2 "no suitable devices found" on Arch Linux

__Expected behavior:__

`tsh fido2 diag` and similar commands work (like `tsh login`).

__Current behavior:__

```shell
$ tsh fido2 diag -d
Tap your security key
            DEBU FIDO2: Selecting devices error:[no suitable devices found] webauthncli/fido2.go:592
            DEBU FIDO2: Selecting devices error:[no suitable devices found] webauthncli/fido2.go:592
            DEBU FIDO2: Selecting devices error:[no suitable devices found] webauthncli/fido2.go:592
            DEBU FIDO2: Selecting devices error:[no suitable devices found] webauthncli/fido2.go:592
            DEBU FIDO2: Selecting devices error:[no suitable devices found] webauthncli/fido2.go:592
            DEBU FIDO2: Selecting devices error:[no suitable devices found] webauthncli/fido2.go:592
^C
FIDO2 available: true
Register successful? false
Login successful? false

ERROR REPORT:
Original Error: *errors.errorString context canceled
Stack Trace:
        /go/src/github.com/gravitational/teleport/lib/auth/webauthncli/fido2.go:596 github.com/gravitational/teleport/lib/auth/webauthncli.findSuitableDevicesOrTimeout
        /go/src/github.com/gravitational/teleport/lib/auth/webauthncli/fido2.go:546 github.com/gravitational/teleport/lib/auth/webauthncli.runOnFIDO2Devices
        /go/src/github.com/gravitational/teleport/lib/auth/webauthncli/fido2.go:444 github.com/gravitational/teleport/lib/auth/webauthncli.fido2Register
        /go/src/github.com/gravitational/teleport/lib/auth/webauthncli/fido2_common.go:95 github.com/gravitational/teleport/lib/auth/webauthncli.FIDO2Register
        /go/src/github.com/gravitational/teleport/lib/auth/webauthncli/fido2_common.go:137 github.com/gravitational/teleport/lib/auth/webauthncli.FIDO2Diag
        /go/src/github.com/gravitational/teleport/tool/tsh/fido2.go:26 main.onFIDO2Diag
        /go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:947 main.Run
        /go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:396 main.main
        /opt/go/src/runtime/proc.go:250 runtime.main
        /opt/go/src/runtime/asm_amd64.s:1571 runtime.goexit
User Message: context canceled
```

In that same system libfido2 1.11.0-1 is installed and `fido2-token` works:

```shell
$ fido2-token -L
/dev/hidraw4: vendor=0x1050, product=0x0407 (Yubico Yubikey 4 OTP+U2F+CCID)
```

Current workaround is to set `TELEPORT_FIDO2=0` in the environment (falls back to the U2F backend).

Thread on Slack: https://goteleport.slack.com/archives/CEZH6UL64/p1658255710172229

__Bug details:__

- Teleport version: 10.0
- Recreation steps: see above
- Debug logs: see above

Alright, so I took another look at the problem with a buddy of mine and we got so far that we built the code on our own. The following build produces a version of `tsh` on my system, with which I can login into the cluster:
```shell
# use dynamicallly linked libfido2
$ make WITH_FIDO=dynamic build/tsh
....

# Weirdly (or maybe thats why its works)for that binary the FIDO2 features are unavailiable
$ ./build/tsh fido2 diag
FIDO2 available: false
Register successful? false
Login successful? false


# unfortunately the static build does not work, so there is no comparison
$ make FIDO2=yes build/tsh  
GOOS=linux GOARCH=amd64 CGO_ENABLED=1 go build -tags " libfido2 libfido2static " -o build/tsh  -ldflags
'-w -s' ./tool/tsh
# pkg-config --cflags  -- libfido2-static
Package libfido2-static was not found in the pkg-config search path.
Perhaps you should add the directory containing `libfido2-static.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libfido2-static', required by 'virtual:world', not found
pkg-config: exit status 1
make: *** [Makefile:239: build/tsh] Error 2
```

Said friend also uses the binary distribution but for him it all works 🤔 Maybe we get to test his yubikey on my machine and the other way around in the next weeks.

--- a/lib/auth/webauthncli/fido2.go
+++ b/lib/auth/webauthncli/fido2.go
@@ -528,6 +528,9 @@ type deviceCallbackFunc func(dev FIDODevice, info *deviceInfo, pin string) error
 // (RegisterPrompt happens to match the minimal interface required.)
 type runPrompt RegisterPrompt
 
+// errNoSuitableDevices is used internally to loop over findSuitableDevices.
+var errNoSuitableDevices = errors.New("no suitable devices found")
+
 func runOnFIDO2Devices(
 	ctx context.Context,
 	prompt runPrompt,
@@ -537,7 +540,7 @@ func runOnFIDO2Devices(
 	knownPaths := make(map[string]struct{}) // filled by findSuitableDevices*
 	prompted := false
 	devices, err := findSuitableDevices(filter, knownPaths)
-	if err != nil {
+	if errors.Is(err, errNoSuitableDevices) {
 		// No readily available devices means we need to prompt, otherwise the
 		// user gets no feedback whatsoever.
 		prompt.PromptTouch()
@@ -585,11 +588,15 @@ func findSuitableDevicesOrTimeout(
 	defer ticker.Stop()
 
 	for {
-		devices, err := findSuitableDevices(filter, knownPaths)
-		if err == nil {
+		switch devices, err := findSuitableDevices(filter, knownPaths); {
+		case err == nil:
 			return devices, nil
+		case errors.Is(err, errNoSuitableDevices):
+			// OK, carry on until we find a device or timeout.
+		default:
+			// Unexpected, abort.
+			return nil, trace.Wrap(err)
 		}
-		log.WithError(err).Debug("FIDO2: Selecting devices")
 
 		select {
 		case <-ctx.Done():
@@ -623,6 +630,11 @@ func findSuitableDevices(filter deviceFilterFunc, knownPaths map[string]struct{}
 		for i := 0; i < infoAttempts; i++ {
 			info, err = dev.Info()
 			switch {
+			case errors.Is(err, libfido2.ErrNotFIDO2):
+				// Use an empty info and carry on.
+				// A FIDO/U2F device has no capabilities beyond MFA
+				// registrations/assertions.
+				info = &libfido2.DeviceInfo{}
 			case errors.Is(err, libfido2.ErrTX):
 				// Happens occasionally, give the device a short grace period and retry.
 				time.Sleep(1 * time.Millisecond)
@@ -649,7 +661,7 @@ func findSuitableDevices(filter deviceFilterFunc, knownPaths map[string]struct{}
 
 	l := len(devs)
 	if l == 0 {
-		return nil, errors.New("no suitable devices found")
+		return nil, errNoSuitableDevices
 	}
 	log.Debugf("FIDO2: Found %v suitable devices", l)
 

--- a/lib/auth/webauthncli/fido2_test.go
+++ b/lib/auth/webauthncli/fido2_test.go
@@ -944,6 +944,63 @@ func TestFIDO2Login_PromptTouch(t *testing.T) {
 	}
 }
 
+func TestFIDO2Login_u2fDevice(t *testing.T) {
+	dev := mustNewFIDO2Device("/u2f", "" /* pin */, nil /* info */)
+	dev.u2fOnly = true
+
+	f2 := newFakeFIDO2(dev).withNonMeteredLocations()
+	f2.setCallbacks()
+
+	const rpID = "example.com"
+	const origin = "https://example.com"
+
+	// Set a ctx timeout in case something goes wrong.
+	// Under normal circumstances the test gets nowhere near this timeout.
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	cc := &wanlib.CredentialCreation{
+		Response: protocol.PublicKeyCredentialCreationOptions{
+			Challenge: []byte{1, 2, 3, 4, 5}, // arbitrary
+			RelyingParty: protocol.RelyingPartyEntity{
+				ID: rpID,
+			},
+			Parameters: []protocol.CredentialParameter{
+				{
+					Type:      protocol.PublicKeyCredentialType,
+					Algorithm: webauthncose.AlgES256,
+				},
+			},
+			AuthenticatorSelection: protocol.AuthenticatorSelection{
+				UserVerification: protocol.VerificationDiscouraged,
+			},
+			Attestation: protocol.PreferNoAttestation,
+		},
+	}
+
+	dev.setUP() // simulate touch
+	ccr, err := wancli.FIDO2Register(ctx, origin, cc, dev /* prompt */)
+	require.NoError(t, err, "FIDO2Register errored")
+
+	assertion := &wanlib.CredentialAssertion{
+		Response: protocol.PublicKeyCredentialRequestOptions{
+			Challenge:      []byte{1, 2, 3, 4, 5}, // arbitrary
+			RelyingPartyID: rpID,
+			AllowedCredentials: []protocol.CredentialDescriptor{
+				{
+					Type:         protocol.PublicKeyCredentialType,
+					CredentialID: ccr.GetWebauthn().GetRawId(),
+				},
+			},
+			UserVerification: protocol.VerificationDiscouraged,
+		},
+	}
+
+	dev.setUP() // simulate touch
+	_, _, err = wancli.FIDO2Login(ctx, origin, assertion, dev /* prompt */, nil /* opts */)
+	assert.NoError(t, err, "FIDO2Login errored")
+}
+
 func TestFIDO2Login_errors(t *testing.T) {
 	resetFIDO2AfterTests(t)
 
@@ -1584,6 +1641,10 @@ type fakeFIDO2Device struct {
 	// conditions.
 	failUV bool
 
+	// Set to true to simulate an U2F-only device.
+	// Causes libfido2.ErrNotFIDO2 on Info.
+	u2fOnly bool
+
 	path        string
 	info        *libfido2.DeviceInfo
 	pin         string
@@ -1658,6 +1719,9 @@ func (f *fakeFIDO2Device) cert() []byte {
 }
 
 func (f *fakeFIDO2Device) Info() (*libfido2.DeviceInfo, error) {
+	if f.u2fOnly {
+		return nil, libfido2.ErrNotFIDO2
+	}
 	return f.info, nil
 }
 
@@ -1695,6 +1759,9 @@ func (f *fakeFIDO2Device) MakeCredential(
 		return nil, libfido2.ErrUnsupportedOption
 	case opts.UV == libfido2.True && !f.hasUV():
 		return nil, libfido2.ErrUnsupportedOption // PIN authenticators don't like UV
+	case opts.RK == libfido2.True && !f.hasRK():
+		// TODO(codingllama): Confirm scenario with a real authenticator.
+		return nil, libfido2.ErrUnsupportedOption
 	}
 
 	// Validate PIN regardless of opts.
@@ -1865,6 +1932,10 @@ func (f *fakeFIDO2Device) hasClientPin() bool {
 	return f.hasBoolOpt("clientPin")
 }
 
+func (f *fakeFIDO2Device) hasRK() bool {
+	return f.hasBoolOpt("rk")
+}
+
 func (f *fakeFIDO2Device) hasUV() bool {
 	return f.hasBoolOpt("uv")
 }
@@ -1874,6 +1945,10 @@ func (f *fakeFIDO2Device) isBio() bool {
 }
 
 func (f *fakeFIDO2Device) hasBoolOpt(name string) bool {
+	if f.info == nil {
+		return false
+	}
+
 	for _, opt := range f.info.Options {
 		if opt.Name == name {
 			return opt.Value == libfido2.True

